File: leetcode/106-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
/*
Given inorder and postorder traversal of a tree, construct the binary tree.

Note:
You may assume that duplicates do not exist in the tree.
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return buildTree(inorder, 0, inorder.size(), postorder, 0, postorder.size());
    }

private:
    TreeNode* buildTree(
            const vector<int> &inorder,
            int ibegin,
            int iend,
            const vector<int> &postorder,
            int pbegin,
            int pend) {
        if (pbegin == pend || ibegin == iend) {
            return nullptr;
        }
        int current = postorder[pend-1];
        auto node = new TreeNode(current);
        int idx = -1;
        for (int i=ibegin; i<iend; ++i) {
            if (inorder[i] == current) {
                idx = i;
                break;
            }
        }
        node->left = buildTree(inorder, ibegin, idx, postorder, pbegin, pbegin+idx-ibegin);
        node->right = buildTree(inorder, idx+1, iend, postorder, pbegin+idx-ibegin, pend-1);

        return node;
    }
};

File: leetcode/106-construct-binary-tree-from-inorder-and-postorder-traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {Integer[]} inorder
# @param {Integer[]} postorder
# @return {TreeNode}
def build_tree(inorder, postorder)
  tree inorder, 0, inorder.size, postorder, postorder.size  
end


def tree(inorder, ib, ie, postorder, pe)
  if ib >= ie or pe <= 0
    nil
  else
    root = TreeNode.new postorder[pe-1]
    inorder_idx = inorder.find_index postorder[pe-1]
    root.right = tree inorder, inorder_idx+1, ie, postorder, pe-1
    root.left = tree inorder, ib, inorder_idx, postorder, pe-1-(ie-inorder_idx-1)
    root
  end
end
